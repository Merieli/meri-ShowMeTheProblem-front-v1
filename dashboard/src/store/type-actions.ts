export enum Actions {
    LOGIN_USER = 'LOGIN_USER',
    REGISTER_USER = 'REGISTER_USER',
    GET_USER = 'GET_USER',

    NEW_FEEDBACK = 'NEW_FEEDBACK',
    GET_ALL_FEEDBACKS = 'GET_ALL_FEEDBACKS',
    GET_FEEDBACKS = 'GET_FEEDBACKS',
    GET_INDEX_FEEDBACK = 'GET_INDEX_FEEDBACK',
    CHANGE_ACTIVE_FEEDBACK = 'CHANGE_ACTIVE_FEEDBACK',
}
