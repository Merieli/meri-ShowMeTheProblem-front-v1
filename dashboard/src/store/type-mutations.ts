export enum Mutations {
    NOTIFY = 'NOTIFY',

    ADD_USER = 'ADD_USER',
    LOGIN_USER = 'LOGIN_USER',
    USER_LOGGED = 'USER_LOGGED',

    ADD_FILTERS = 'ADD_FILTERS',
    ADD_CONFIGURED_FILTERS = 'ADD_CONFIGURED_FILTERS',
    ADD_FEEDBACKS = 'ADD_FEEDBACKS',

    TOOGLE_LOADING = 'TOOGLE_LOADING',
}
