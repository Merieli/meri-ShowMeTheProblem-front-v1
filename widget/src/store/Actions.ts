export enum Actions {
    SAVE_FEEDBACK = 'SAVE_FEEDBACK',
}
