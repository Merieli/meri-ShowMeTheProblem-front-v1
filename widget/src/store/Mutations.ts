export enum Mutations {
    SET_CURRENT_COMPONENT = 'SET_CURRENT_COMPONENT',
    SET_MESSAGE = 'SET_MESSAGE',
    SET_FEEDBACK_TYPE = 'SET_FEEDBACK_TYPE',
    SET_CURRENT_PAGE = 'SET_CURRENT_PAGE',
    SET_APIKEY = 'SET_APIKEY',
    SET_FINGERPRINT = 'SET_FINGERPRINT',
    RESET_STORE = 'RESET_STORE',
}
