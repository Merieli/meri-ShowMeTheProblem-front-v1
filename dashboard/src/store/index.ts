import useNotifier from '../hooks/notifier'
import router from '../router/index'
import { Actions } from './type-actions'
import { Mutations } from './type-mutations'
import callApiClient from '@/http-client/index'
import { TypeOfNotification } from '@/interfaces/INotification'
import { INotification } from '@/interfaces/INotification'
import IUser from '@/interfaces/IUser'
import IUserLogged from '@/interfaces/IUserLogged'
import { createStore } from 'vuex'

export interface EstadoStore {
    users: IUser[]
    notifications: INotification[]
    isLogged: boolean
    userLogged: IUserLogged
}

export const store = createStore<EstadoStore>({
    state: {
        users: [],
        notifications: [],
        isLogged: false,
        userLogged: {
            name: '',
            token: '',
            apiKey: '',
        },
    },
    getters: {},
    mutations: {
        [Mutations.NOTIFY](state, newNotification: INotification) {
            newNotification.id = new Date().getTime()
            state.notifications.push(newNotification)
            setTimeout(() => {
                state.notifications = state.notifications.filter((notif) => notif.id != newNotification.id)
            }, 5000)
        },
        [Mutations.ADD_USER](state, user: IUser) {
            state.users.push(user)
        },
        [Mutations.LOGIN_USER](state, token) {
            state.userLogged.token = token
            state.isLogged = true
        },
        [Mutations.USER_LOGGED](state, data) {
            state.userLogged.name = data.name
            state.userLogged.apiKey = data.apiKey
        },
    },
    actions: {
        /**
         * @name REGISTER_USER
         * @descripton efetua o registro do usuário e o adiciona a store
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.REGISTER_USER]({ commit }, user: IUser) {
            try {
                const response = await callApiClient.user.register(user.name, user.email, user.password)
                commit(Mutations.ADD_USER, response)
            } catch (e) {
                throw new Error(`Não foi possível efetuar cadastro. Confira os dados informados`)
            }
        },

        /**
         * @name LOGIN_USER
         * @description Efetua login do usuário, salva o token na store,
         * redireciona para a página de 'feedbacks' e aciona a action
         * para pegar os dados do usuário logado.
         * @param {IUser} user usuário com name, email e password
         */
        async [Actions.LOGIN_USER]({ commit, dispatch, state }, user: IUser) {
            try {
                const response = await callApiClient.user.login(user.email, user.password)
                commit(Mutations.LOGIN_USER, response.token)
                router.push('/feedbacks')
                dispatch(Actions.GET_USER, state.userLogged.token)
            } catch (e) {
                const { notify } = useNotifier()
                notify(TypeOfNotification.FALHA, 'Não foi possível efetuar login', 'Confira os dados informados.')
                throw new Error(`Não foi possível efetuar login. Confira os dados informados`)
            }
        },

        /**
         * @name GET_USER
         * @descripton Captura os dados do usuários e salva na store
         * @param token chave de acesso do usuário
         */
        async [Actions.GET_USER]({ commit }, token) {
            try {
                const response = await callApiClient.user.show(token)
                commit(Mutations.USER_LOGGED, response)
            } catch (e) {
                throw new Error(`Não foi possível capturar os dados do usuário.`)
            }
        },
    },
    modules: {},
})
