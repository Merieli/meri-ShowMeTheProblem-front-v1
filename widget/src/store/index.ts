import httpClient from '@/http';
import { GettersStore, MutationsStore, StateStore } from '@/interfaces/';
import { ActionsStore } from '@/interfaces/ActionsStore';
import { InjectionKey } from 'vue';
import { createStore, Store, useStore as baseUseStore } from 'vuex';
import { Actions } from './Actions';
import { Mutations } from './Mutations';

export const state: StateStore = {
    currentComponent: 'SelectFeedbackType',
    message: '',
    feedbackType: '',
    fingerprint: '',
    apiKey: '',
    currentPage: '',
};

export const mutations: MutationsStore = {
    [Mutations.SET_CURRENT_COMPONENT](state, componentName): void {
        state.currentComponent = componentName;
    },
    [Mutations.SET_MESSAGE](state, message): void {
        state.message = message;
    },
    [Mutations.SET_FEEDBACK_TYPE](state, type): void {
        state.feedbackType = type;
    },
    [Mutations.SET_CURRENT_PAGE](state, page): void {
        state.currentPage = page;
    },
    [Mutations.SET_APIKEY](state, apiKey): void {
        state.apiKey = apiKey;
    },
    [Mutations.SET_FINGERPRINT](state, fingerprint): void {
        state.fingerprint = fingerprint;
    },
    [Mutations.RESET_STORE](state): void {
        state.currentComponent = 'SelectFeedbackType';
        state.message = '';
        state.feedbackType = '';
        state.currentPage = '';
        state.apiKey = '';
        state.fingerprint = '';
    },
};

export const getters: GettersStore = {
    currentComponent(state): string {
        return state.currentComponent;
    },
    feedbackType(state): string {
        return state.feedbackType;
    },
};

export const actions: ActionsStore = {
    async [Actions.SAVE_FEEDBACK]({ state, commit }, feedback): Promise<void> {
        try {
            commit(Mutations.SET_MESSAGE, feedback);
            await httpClient.feedbacks.create({
                type: state.feedbackType,
                text: state.message,
                page: state.currentPage,
                apiKey: state.apiKey,
                device: window.navigator.userAgent,
                fingerprint: state.fingerprint,
            });
            commit(Mutations.SET_CURRENT_COMPONENT, 'SuccessSave');
        } catch (error) {
            commit(Mutations.SET_CURRENT_COMPONENT, 'ErrorSave');
            throw new Error('Não foi possível salvar o feedback!');
        }
    },
};

export const store = createStore({
    state,
    getters,
    mutations,
    actions,
    modules: {},
});

export const key: InjectionKey<Store<StateStore>> = Symbol();

export function useStore(): Store<StateStore> {
    return baseUseStore(key);
}
