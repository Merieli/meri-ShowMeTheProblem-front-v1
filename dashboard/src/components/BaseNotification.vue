<template>
    <div class="notification w-80 rounded-md" id="notification">
        <div
            class="rounded-md p-4 m-4"
            v-for="notication in notifications"
            :key="notication.id"
            :class="classAlert[notication.type]"
        >
            <h6 class="notification__header font-bold" data-test="title">{{ notication.title }}</h6>
            <p class="notification__body font-normal">{{ notication.text }}</p>
        </div>
    </div>
</template>

<script lang="ts">
import { TypeOfNotification } from '../interfaces'
import { useStore } from '../store'
import { defineComponent, computed } from 'vue'

export default defineComponent({
    name: 'BaseNotification',
    data() {
        return {
            classAlert: {
                [TypeOfNotification.SUCESSO]: 'notification__success',
                [TypeOfNotification.ATENCAO]: 'notification__warning',
                [TypeOfNotification.FALHA]: 'notification__danger',
            },
        }
    },

    setup() {
        const store = useStore()

        return {
            notifications: computed(() => store.getters.notifications),
        }
    },
})
</script>

<style scoped lang="scss">
.notification {
    position: absolute;
    right: 0;
    width: 300px;
    padding: 8px;
    z-index: 10;
    top: 0;

    &__success {
        background-color: rgb(247 254 231);
        color: rgb(77 124 15);
    }

    &__warning {
        color: rgb(234 88 12);
        background-color: rgb(255 237 213);
    }

    &__danger {
        color: rgb(220 38 38);
        background-color: rgb(254 226 226);
    }
}
</style>
