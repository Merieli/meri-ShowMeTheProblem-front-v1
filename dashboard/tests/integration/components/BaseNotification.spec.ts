import BaseNotification from '@/components/BaseNotification.vue'
import { key } from '@/store'
import { mount } from '@vue/test-utils'
import { createStore } from 'vuex'

describe('BaseNotification', () => {
    const mockNotifications = () => mockState.notifications
    const mockState = {
        notifications: [
            {
                type: 'SUCESSO',
                title: 'Teste',
                text: 'Uma notificação exibida',
                id: 12,
            },
        ],
    }
    let mockVuexStore: any

    beforeEach(() => {
        mockVuexStore = createStore({
            state: mockState,
            getters: { notifications: mockNotifications },
        })
    })

    function factory(storeMock: any) {
        return mount(BaseNotification, {
            global: {
                plugins: [[storeMock, key]],
            },
        })
    }

    describe('Integração/Componente', () => {
        describe('Renderização:', () => {
            test('Dado o html quando renderizado então deve ter os mesmos dados do snapshot gravado', () => {
                const wrapper = factory(mockVuexStore)
                expect(wrapper.html()).toMatchSnapshot()
            })

            test('Dado a store quando tiver dados no state de notifications então o título da notificação deve ser "Teste"', async () => {
                const wrapper = factory(mockVuexStore)

                expect(wrapper.find('[data-test="title"]').text().trim()).toEqual('Teste')
                expect(wrapper.html()).toContain('Teste')
            })
        })

        describe('Comportamento:', () => {
            return
        })
    })
})
