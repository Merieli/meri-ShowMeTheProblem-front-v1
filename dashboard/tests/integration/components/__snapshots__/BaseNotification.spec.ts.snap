// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseNotification Integração/Componente Renderização: Dado o html quando renderizado então deve ter os mesmos dados do snapshot gravado 1`] = `
"<div class=\\"notification w-80 rounded-md\\" id=\\"notification\\">
  <div class=\\"rounded-md p-4 m-4\\">
    <h6 class=\\"notification__header font-bold\\" data-test=\\"title\\">Teste</h6>
    <p class=\\"notification__body font-normal\\">Uma notificação exibida</p>
  </div>
</div>"
`;
